data/shared/citizen/scripting/v8/msgpack.js

```diff
@@ -1,30 +1,2331 @@
+					else {
+						let i = src[position$1++];
+						if ((i & 0x80) > 0) {
+							position$1 -= 9;
+							return
+						}
+						return fromCharCode(a, b, c, d, e, f, g, h, i)
+					}
+				} else if (length < 12) {
+					let i = src[position$1++];
+					let j = src[position$1++];
+					if ((i & 0x80) > 0 || (j & 0x80) > 0) {
+						position$1 -= 10;
+						return
+					}
+					if (length < 11)
+						return fromCharCode(a, b, c, d, e, f, g, h, i, j)
+					let k = src[position$1++];
+					if ((k & 0x80) > 0) {
+						position$1 -= 11;
+						return
+					}
+					return fromCharCode(a, b, c, d, e, f, g, h, i, j, k)
+				} else {
+					let i = src[position$1++];
+					let j = src[position$1++];
+					let k = src[position$1++];
+					let l = src[position$1++];
+					if ((i & 0x80) > 0 || (j & 0x80) > 0 || (k & 0x80) > 0 || (l & 0x80) > 0) {
+						position$1 -= 12;
+						return
+					}
+					if (length < 14) {
+						if (length === 12)
+							return fromCharCode(a, b, c, d, e, f, g, h, i, j, k, l)
+						else {
+							let m = src[position$1++];
+							if ((m & 0x80) > 0) {
+								position$1 -= 13;
+								return
+							}
+							return fromCharCode(a, b, c, d, e, f, g, h, i, j, k, l, m)
+						}
+					} else {
+						let m = src[position$1++];
+						let n = src[position$1++];
+						if ((m & 0x80) > 0 || (n & 0x80) > 0) {
+							position$1 -= 14;
+							return
+						}
+						if (length < 15)
+							return fromCharCode(a, b, c, d, e, f, g, h, i, j, k, l, m, n)
+						let o = src[position$1++];
+						if ((o & 0x80) > 0) {
+							position$1 -= 15;
+							return
+						}
+						return fromCharCode(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o)
+					}
+				}
+			}
+		}
+	}
+
+	function readOnlyJSString() {
+		let token = src[position$1++];
+		let length;
+		if (token < 0xc0) {
+			// fixstr
+			length = token - 0xa0;
+		} else {
+			switch(token) {
+				case 0xd9:
+				// str 8
+					length = src[position$1++];
+					break
+				case 0xda:
+				// str 16
+					length = dataView.getUint16(position$1);
+					position$1 += 2;
+					break
+				case 0xdb:
+				// str 32
+					length = dataView.getUint32(position$1);
+					position$1 += 4;
+					break
+				default:
+					throw new Error('Expected string')
+			}
+		}
+		return readStringJS(length)
+	}
+
+
+	function readBin(length) {
+		return currentUnpackr.copyBuffers ?
+			// specifically use the copying slice (not the node one)
+			Uint8Array.prototype.slice.call(src, position$1, position$1 += length) :
+			src.subarray(position$1, position$1 += length)
+	}
+	function readExt(length) {
+		let type = src[position$1++];
+		if (currentExtensions[type]) {
+			let end;
+			return currentExtensions[type](src.subarray(position$1, end = (position$1 += length)), (readPosition) => {
+				position$1 = readPosition;
+				try {
+					return read();
+				} finally {
+					position$1 = end;
+				}
+			})
+		}
+		else
+			throw new Error('Unknown extension type ' + type)

```

Comment:
fwiw this will break compatibility for any type that has no extension. For example, unpacking a vector2 from Lua results in

```js
return {
  "buffer": {
    "type": "Buffer",
    "data": number[] // i.e. [11,205,143,63,183,230,7,64]
  },
  "type": 20
}
```

Reply:
We should probably just handle the conversions here anyways, since sending vectors to/from scrts is currently very hacky and I doubt may people do it currently.

Reply:
Created an array containing 100 objects and another with 1000 objects. Performed 1000 iterations.

msgpack-lite
```
Running benchmark for packing...
Pack - Min: 0.0568 ms, Max: 1.1691 ms, Avg: 0.080365 ms
Running benchmark for unpacking...
Unpack - Min: 0.0482 ms, Max: 1.1444 ms, Avg: 0.061173 ms

Running benchmark for packing...
Pack - Min: 5.2618 ms, Max: 13.8522 ms, Avg: 5.99034 ms
Running benchmark for unpacking...
Unpack - Min: 5.0637 ms, Max: 20.6582 ms, Avg: 5.366889 ms
```

msgpackr
```
Running benchmark for packing...
Pack - Min: 0.0185 ms, Max: 4.2132 ms, Avg: 0.03549 ms
Running benchmark for unpacking...
Unpack - Min: 0.035 ms, Max: 1.9161 ms, Avg: 0.059167 ms

Running benchmark for packing...
Pack - Min: 1.8215 ms, Max: 10.3579 ms, Avg: 1.995244 ms
Running benchmark for unpacking...
Unpack - Min: 3.4534 ms, Max: 18.5718 ms, Avg: 3.717805 ms
```